redash/models/changes.py

```diff
@@ -46,48 +54,85 @@ def last_change(cls, obj):
+                result[col] = change
+
+        if reset:
+            changes.clear()
+
+    return result
+
+
+def _patch_setattr_method(cls, attributes):
+    original_setattr = cls.__setattr__
+
+    def new_setattr(self, key, value):
+        if key in attributes:
+            if not hasattr(self, "__changes__"):
+                self.__changes__ = {}
+            change = self.__changes__.get(key, {
+                "previous": getattr(self, key),
+                "current": None,
+            })
+            change["current"] = value
+            self.__changes__[key] = change
+
+        original_setattr(self, key, value)
+
+    cls.__setattr__ = new_setattr
+
+
+def _patch_record_changes_method(cls, parent_attr):
+    def record_changes(self, changed_by, change_type=Change.Type.Modified):
+        session = object_session(self)
+        if not session:
+            return
+
+        changes = _get_object_changes(self)
+        # for `created` and `deleted` log even empty changes set
+        if not changes and (change_type == Change.Type.Modified):
+            return
+
+        changes = Change(
+            object=self,
+            object_version=getattr(self, "version", None),
+            user=changed_by,
+            change={
+                "object_type": self.__table__.name,
+                "object_id": self.id,
+                "change_type": change_type,
+                "changes": changes,
+            },
+        )
 
-    def __setattr__(self, key, value):
-        if self._clean_values is None:
-            self.prep_cleanvalues()
-        for attr in inspect(self.__class__).column_attrs:
-            col, = attr.columns
-            previous = getattr(self, attr.key, None)
-            self._clean_values[col.name] = previous
+        if parent_attr:
+            changes.object = getattr(self, parent_attr, self)
+            changes.object_version = getattr(changes.object, "version", None),
 
-        super(ChangeTrackingMixin, self).__setattr__(key, value)
+        session.add(changes)
 
-    def record_changes(self, changed_by):
-        db.session.add(self)
-        db.session.flush()
-        changes = {}
-        for attr in inspect(self.__class__).column_attrs:
-            col, = attr.columns
-            if attr.key not in self.skipped_fields:
-                changes[col.name] = {
-                    "previous": self._clean_values[col.name],
-                    "current": getattr(self, attr.key),
-                }
-
-        db.session.add(
-            Change(
-                object=self,
-                object_version=self.version,
-                user=changed_by,
-                change=changes,
-            )
-        )
+    cls.record_changes = record_changes
+
+
+def track_changes(attributes, parent_attr=None):
+    attributes = set(attributes) - {"id", "created_at", "updated_at", "version"}
+
+    # monkey-patch class because inheritance will break SQLAlchemy
```

Comment:
I decided to create a decorator for change tracking because it looks nice when you use in when defining a model. Decorator needs to modify original model class: intercept `__setattr__` and add a method to save changes. First I tried to inherit from original model class (`class ChangeTracking(cls)`) and it even worked. But in some cases SQLA was throwing errors - something like `class is not mapped` or like this (in particular that happened in tests). So eventually I checked how they implement own decorators in SQLA (e.g. `sqlalchemy_utils.models.generic_repr`) and realized that instead of using inheritance, they just monkey-patch the class (replace existing methods with own ones). I tried to do so - and it worked. OOP in Python is a bit weird; I guess when I inherited my class - it changed something in a mechanism SQLA uses to get model's metadata; when patching class I in fact keep the class itself untouched, and everything works